packages/adblocker/src/engine/engine.ts

```diff
@@ -1499,13 +1527,115 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
+              }
+            }
+            const removeparamIgnoreFilter: NetworkFilter | undefined =
+              // `result.exception` is conditionally matched only if `result.filter` is available.
+              (result.filter === undefined
+                ? this.exceptions.match(request, this.isFilterExcluded.bind(this))
+                : result.exception) || removeparamExceptions.get('');
+
+            for (const [key, filter] of removeparamFilters) {
+              // Remove all params in case of option value is empty.
+              // We will not match individual exceptions since it has a higher priority than them.
+              if (key === '') {
+                // In case of non-existence of global exception, we will remove all params.
+                if (removeparamIgnoreFilter === undefined) {
+                  // We need to collect all keys before the execution of `delete()`.
+                  // Running `delete()` will inference with an iterator and its inner index.
+                  for (const key of Array.from(searchParams.keys())) {
+                    searchParams.delete(key);
+                  }
+                  modified = true;
+                }
+                this.emit(
+                  'filter-matched',
+                  { filter, exception: removeparamIgnoreFilter },
+                  { request, filterType: FilterType.NETWORK },
+                );
+                break;
+              }
+
+              if (
+                !key.startsWith('~') &&
+                // Try to find `?${key}` pattern.
+                searchParamLiteral.slice(1, key.length + 1) !== key &&
+                !searchParamLiteral.includes(`&${key}`)
+              ) {
+                continue;
+              }
+
+              const exception = removeparamExceptions.get(key) ?? removeparamIgnoreFilter;
+              if (exception === undefined) {
+                // Handle removeparam inversions.
+                if (key.startsWith('~')) {
+                  const inversionKey = key.slice(1);
+                  for (const param of Array.from(searchParams.keys())) {
+                    if (param !== inversionKey && !removeparamExceptions.has(param)) {
+                      searchParams.delete(param);
+                      modified = true;
+                    }
+                  }
+                } else {
+                  searchParams.delete(key);
+                  modified = true;
+                }
+              }
+              this.emit(
+                'filter-matched',
+                { filter, exception },
+                { request, filterType: FilterType.NETWORK },
+              );
+            }
+
+            let url: string | undefined;
+            if (modified) {
+              url = request.url.slice(0, searchParamSeparatorIndex);
+              if (searchParams.size > 0) {
+                url += '?' + searchParams.toString();
+              }
+            }
+
+            result.rewrite = {
```

Comment:
Should we set this only when `modified` above is `true`?

Reply:
fixed in 8f87db81a3564ae1c59c4ea934a99c3e25cbec0a